fs/nfs/nfs4proc.c

```diff
@@ -7771,6 +7775,475 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
+ * functions in nfsd.
+ */
+
+/*
+ * return the size of the struct nfs4_acl required to represent an acl
+ * with @entries entries.
+ */
+static int nfs4_acl_bytes(int entries)
+{
+	return sizeof(struct nfs4_acl) + entries * sizeof(struct nfs4_ace);
+}
+
+static struct {
+	char *string;
+	int   stringlen;
+	int type;
+} s2t_map[] = {
+	{
+		.string    = "OWNER@",
+		.stringlen = sizeof("OWNER@") - 1,
+		.type      = NFS4_ACL_WHO_OWNER,
+	},
+	{
+		.string    = "GROUP@",
+		.stringlen = sizeof("GROUP@") - 1,
+		.type      = NFS4_ACL_WHO_GROUP,
+	},
+	{
+		.string    = "EVERYONE@",
+		.stringlen = sizeof("EVERYONE@") - 1,
+		.type      = NFS4_ACL_WHO_EVERYONE,
+	},
+};
+
+static int nfs4_acl_get_whotype(char *p, u32 len)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(s2t_map); i++) {
+		if (s2t_map[i].stringlen == len &&
+		    (memcmp(s2t_map[i].string, p, len) == 0))
+			return s2t_map[i].type;
+	}
+	return NFS4_ACL_WHO_NAMED;
+}
+
+/*
+ * XDR encode the ace who into the supplied buffer.
+ *
+ * Return the number of bytes (including the leading nbytes and
+ * any padding necessary)
+ */
+static int nfs4_ace_encode_who(struct nfs4_ace *ace, char *buf)
+{
+	u32 *p = (u32 *)buf;
+	int count = -EINVAL;
+	const char *pstr = NULL;
+	char idbuf[11];
+	int total_bytes, pad_count;
+
+	/*
+	 * First find the size of the string (into count)
+	 *
+	 * Will also have some side-effects which will be used below:
+	 * either idbuf will be populated (for NFS4_ACL_WHO_NAMED) or
+	 * pstr will point into s2t_map.
+	 */
+	if (ace->whotype == NFS4_ACL_WHO_NAMED) {
+		u32 id;
+
+		if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP)
+			id = from_kgid_munged(&init_user_ns, ace->who_gid);
+		else
+			id = from_kuid_munged(&init_user_ns, ace->who_uid);
+		sprintf(idbuf, "%u", id);
+		count = strlen(idbuf);
+	} else {
+		int i;
+
+		for (i = 0; i < ARRAY_SIZE(s2t_map); i++) {
+			if (ace->whotype == s2t_map[i].type) {
+				count = s2t_map[i].stringlen;
+				pstr = s2t_map[i].string;
+				break;
+			}
+		}
+	}
+	if (count < 0)
+		return count; /* error */
+
+	/*
+	 * who is a utf8str_mixed, i.e. a utf8string.  Per RFC 1832
+	 * this is encoded as a uint32_t length followed by
+	 * string bytes, and then padded by null up to a 4 byte
+	 * boundary.
+	 */
+	total_bytes = 4 + (XDR_QUADLEN(count) << 2);
+	if (!buf)
+		return total_bytes;
+	pad_count = total_bytes - (4 + count);
+
+	/*
+	 * A buffer was supplied. Write the string length, followed by the
+	 * string bytes and then null padding, if necessary.
+	 */
+	*p++ = htonl(count);
+	if (ace->whotype == NFS4_ACL_WHO_NAMED)
+		memcpy(p, idbuf, count);
+	else
+		memcpy(p, pstr, count);
+	if (pad_count) {
+		char *pad = buf + (4 + count);
+
+		memset(pad, 0, pad_count);
+	}
+	return total_bytes;
+}
+
+/*
+ * Convert the specified (numeric) name into a u32 ID.
+ *
+ * Returns true on success.
+ */
+static bool numeric_name_to_id(const char *name, u32 namelen, u32 *id)
+{
+	int ret;
+	char buf[11];
+
+	if (namelen + 1 > sizeof(buf))
```

Comment:
Should we also add a NULL check for `id`?

Reply:
BTW, also made 64bit resilient (to future-proof, based on possible changes wrt `uid_t` / `gid_t`)